### ISSUE TITLE
MQTT trigger node executes more than one time

### ISSUE THREAD

**Issue by edlundin** [n8n-io/n8n#3824]
Hi!

After encountering significant problems with the MQTT trigger node, I decided to write this up, hoping it could help identify the origin of the issue.

**Describe the bug**
The MQTT trigger node behaviour is erratic.

I wrote two workflows, each containing an MQTT trigger node and a function one (it only logs "Done!" in the console). Both of the MQTT nodes are subscribing to the same topic.

It turns out that sometimes my workflow is executed multiple times, even when receiving only one message.

When activating only one of the two flows, receiving a message on the subscribed topic causes a spike in resource consumption. Sometimes it consumes a lot of memory (from ~400MB to ~3.2GB, for one message in the worst case I witnessed), sometimes not so much, depending on how many times the workflow ran.

The real issue occurs when enabling both workflows.
Like above, the execution history shows multiple executions for one received message. Besides both of them being enabled (MQTT 1 & 2), only one is executed (MQTT 2).

<img width="1323" alt="execution_history" src="https://user-images.githubusercontent.com/68131703/182623101-b83352e1-bbfb-41dd-82dc-6a8d9f5e2d7c.png">

The memory consumption explodes, as well as the system charge. Often, it will fill the RAM and swap until the whole machine crashes. Even if you somehow survive this, the following message often triggers the same bug, even if one of the workflows has been disabled.

A screenshot of HTOP a bit before the VM dies:

<img width="1054" alt="Screenshot 2022-08-03 at 15 45 10" src="https://user-images.githubusercontent.com/68131703/182624837-83c78a56-0ed4-46cc-91f5-b9acce5c90ce.png">

**To Reproduce**
Steps to reproduce the behaviour:
1. Create a workflow with an MQTT trigger node and a function node
2. Configure the MQTT trigger node
3. Duplicate the workflow
4. Enable both of the workflows
5. Publish messages on the subscribed topic

**Expected behaviour**
Only one execution per workflow per MQTT message.
Memory consumption consistency for this particular node.

**Environment:**
 - OS: Debian 11.4 / n8n docker image
 - n8n Version: 0.188.0
 - Node.js Version: node v18.7.0 / node v16.16
 - Database system: SQLite
 - Operation mode: Trigger
 - VM configuration: 4 cores - 8GB of RAM

**Additional context**
Sent MQTT message, always in QOS 0:

```JSON
{
  "msg": "hello2"
}
```

**Comment by rvalle** [comment]
I am seeing this problem too... multiple executions for the same message. I have verified that it is actually the same payload triggering multiple times the same workflow.

I connect other software, such as the MQTT explorer, and they see the message only once, yet in n8n, multiple executions are triggered.

**Comment by rvalle** [comment]
Actually this is my second MQTT workflow, but I have looked but at executions of the previous one, and I can also see 10s of executions of the workflow within one second.

This may be a regression, because I have had my other workflow running for about a year, and back then, when I set it up, I did not notice any issue like this. I would have noticed this.

**Comment by rvalle** [comment]
cc @Joffcom MQTT seems to be complete broken to me, and affects the stability of n8n. Once incoming message triggers 10s of executions.

I am reviewing the implementation of this trigger and looking into MQTT.js for clues of what could be going wrong.

Any idea of what could be happening?

I will check if I can reproduce this behavior from a vanilla node 16 test.

**Comment by rvalle** [comment]
Here is how dramatic the resource leakage by MQTT trigger is:

![n8n MQTT - Grafana](https://user-images.githubusercontent.com/412837/186736318-6975a4cb-44e1-48a0-a3a2-fe2bcfb8497f.png)

you can guess the blue n8n container jumped to 5GB almost and crashed its VM.

**Comment by rvalle** [comment]
I did some additional testing with a brand new container, and I got 2 listeners to work normally, and adding a third reproduced the problem. So, @edlundin seems right, the cardinality of listeners seems to play a important role.

**Comment by rvalle** [comment]
I think I found the problem.

@edlundin did you fill the client id in your MQTT listener settings?

I did. And that causes the trouble. The issue is that according to the MQTT specifications the broker should not allow more than one connection with the same client ID (not to confuse with username/password). 

If you create several listeners with the same client id it will crash, but if you leave it blank, n8n seems to generate a different client id per listener.

The fact is that client-id should not be a connection setting but actually a listener or publisher option, and the recommendation should be to leave it blank for auto-generation or be warned that must be assured uniqueness (n8n could also enforce uniqueness).

During the testing I also noticed that MQTT listeners whose connections are closed during n8n start will crash the container and put it in an state in which it cannot be fixed. A safe start mode would be nice, maybe it exists, so that you can disable "poisoned" workflows.

So, I have so far setup MQTT listeners, with same credentials and black client id and seems to be working now.

I use mosquito-mqtt as broker. 

cc @Joffcom

**Comment by edlundin** [comment]
I checked my client ID, and it seems to be the issue's root since they were the same in both listeners. To avoid any confusion in the future, maybe the clientId field should be outside of MQTT credentials.

I am using EMQX as a broker.
I tried using an external MQTT client (MQTTX) with the same client ID as the listener in n8n. It seems that (in the case of EMQX at least), the last client initiating a connection with a conflicting client ID will get to keep it, while the server will disconnect the already connected client.

This client will then issue a reconnect, always rejected by the server. Could it be that n8n as a problem with the state of this last client?

**Comment by rvalle** [comment]
I think multiple connections with same client-id are basically banned by the protocol. Like 2 computers with the same IP on the same network, anything can go wrong from then on.

I think the but is related to UX, client-id should not be asked when setting credentials because it creates the error. I did set n8n as client id.

client id should be generated automatically without asking the user, and perhaps be an advanced option of the listener or published in case there is a particular use case that requires specific client-id.

**Comment by Joffcom** [comment]
That is some nice work there @rvalle, I am not sure if we had started to look into this one yet but it sounds like a quick win could be to mark it as advanced with a note about it.

**Comment by edlundin** [comment]
I agree with not asking for the client ID while setting the credentials since multiple listeners can use them, unlike the client ID.
On the other hand, I would not suggest generating the client ID without the user's say. You might want to pre-fill the client ID field with something (e.g. "n8n_$TIMESTAMP_MS"), but ultimately the user has to be able to change it with ease (so without hiding it in an "advanced panel" or equivalent).

Finally, while the protocol prohibits multiple connections with the same client ID, I still think there is a deeper issue in how the MQTT node (and possibly n8n) deals with that error internally. Even if pre-filling the client ID will mitigate the problem, n8n should not crash an entire system because a user forgot or is unaware of the MQTT specification.

**Comment by rvalle** [comment]
@edlundin I mostly agree.
The client ID seems of little value, other than debugging. I have checked other software and all seem to do the same, ask for it and if not autogenerate. n8n is generating mqttjs-ramdomid, so I guess it is the underlying library.

Your proposal of generating for the user to see: n8n-ramdomid is better, because it aids debugging in case you want to actually know which listener is which connection. However that would require first move the client-id as a listener/publisher option.

I agree with @Joffcom that the ultra-quick fix is just to warn users not to set it unless they know the implications (advanced). Currently the implication is that one set of credentials + client id needs to be create per MQTT node. 

In a second step I would move it out of credentials into the nodes with your suggestion of pre-generating it and a warning that replacing it is advanced.

it would be amazing if n8n enforces uniqueness of client-ids, and fails at UI to setup a repeated client-id , but hat is a stretch goal!

**Comment by rvalle** [comment]
@Joffcom I have now been following this for some weeks.

Even thou stability of MQTT listeners improves massively, there remain issues.

In my case listeners will start delivering one extra copy of the message per execution week. i.e. After 2 weeks up and running messages are coming  x3 times. Otherwise the system is stable, no  massive resource leak, etc. A restart of n8n container will fix the issue.

@edlundin are you seeing this behavior too?

**Comment by Joffcom** [comment]
Hey @rvalle,

When you say the listeners are delivering an extra message is that the trigger in n8n receiving an extra message or something else sending a message?

I will have a look later this week to see if I can reproduce the duplicate items being read issue.

**Comment by rvalle** [comment]
@Joffcom I mean that n8n will start duplicating message deliveries. 

I connect n8n and mqtt-explorer to mosquito. n8n consistently reports duplicated messages while mqtt explorer reports only 1 copy. Both systems have connected with the same QOS. And as said, restarting n8n will fix the issue.

I believe the extra messages are been generated by the MQTT stack inside n8n. It will be difficult to reproduce, or will require some kind of stress testing because I only see this condition after one week of n8n running. 

To give you an idea of traffic/size: I have around 40 devices in my MQTT-IOT network, 3 of then trigger workflows. So, in n8n I have 1 MQTT settings, 3 listeners. My actual workflow will send a message to Matrix IM after checking MQTT payload for some condition.

Other projects using this MQTT stack must have experience similar issues. I guess.

**Comment by Joffcom** [comment]
Hey @rvalle,

I did see one person mention duplicate messages on the mqtt.js issue tracker but that was it. In theory if it happens after a week I should be able to set up an MQTT server and fire some messages in and just watch the results to see what happens.

It looks like mqtt.js has a debug env option so I can see if that works as well, When you are running is n8n or mqtt ever restarted?

**Comment by Joffcom** [comment]
Hey @rvalle,

What do you have the QOS option set to? When you say you have 3 listeners is that 3 trigger nodes looking at different topics or the same topic?

**Comment by rvalle** [comment]
@Joffcom  they look at different topics. No special options setup (JSON Parsing only), no QOS setup (defaults to 0).

**Comment by Joffcom** [comment]
@rvalle perfect and how many messages is it roughly a day?

**Comment by rvalle** [comment]
@Joffcom I think roughly around 100 per minute, that would be around 150K per day, and that means that the issue is popping after 1M message approx.

Of course this is for all 40 devices in the MQTT network, perhaps it is only the target messages that matter. Then it may be much less, as only 3 of the 40 devices are involved.

**Comment by Joffcom** [comment]
@rvalle cheers, I realised my testing was a bit too simple so just have the bombing script a go and killed my n8n instance so I will need to bump up the memory a bit on that but I suspect this may come down to resources and something odd in the package we are using.

I don't think this is going to be a quick fix at the moment but I have created N8N-4647 which is our internal reference for tracking this one.

**Comment by rvalle** [comment]
@Joffcom I can surely try, it but I doubt it will work, because when I connect, in parallel MQTT Explorer with a similar QOS it gets only one copy.

But you never know, does not cost anything to try. Will setup one of the listeners with QOS 2 and see the impact.

On another note, I can confirm that it keeps happening after restarts, I was also thinking that I could try to enable remote debugging on my container and connect once the bug is active to open a debug session and try to gain some more insights.

**Comment by kevinrademan** [comment]
I am seeing exactly the same behavior with the redis trigger, so this might not be a MQTT only issue.

**Comment by Joffcom** [comment]
@kevinrademan That is interesting, So are you seeing the same message being pulled multiple times? I would be surprised if it is the same cause but you never know.

**Comment by kevinrademan** [comment]
@Joffcom I would'nt call it pulled multiple times. Its almost like the redis trigger doesnt cleanup some kind of queue resulting in previously executions of the trigger running again. So over time the amount of jobs that gets executed gets more and more.

Heres my usecase. Im doing some local development on a shopify integration and required the webhooks to reach my machine. We have a self hosted version of n8n which recieves the shopify webhook, and just publishes the raw data into a redis queue.

Locally I also have n8n running which uses the redis trigger to receive messages published to a specific channel and then uses the http request node to send along the data to my local dev env.

When I check the execution logs for the publically accessible n8n instance, I can see that 1 shopify webhook request results in 1 entry into the execution history. 

Locally however that 1 execution causes up to 20 executions of my workflow. It also seems to **get worse over time** which makes me think this is some queue or event listener thats not getting cleaned up. It however only executes the 20 jobs when a shopify webhook comes it. So its 1 shopify webhook trigger that causes the 20 jobs to execute. And for the record it doesnt start at 20 jobs. Initially it works and just executes fine, but as time goes along it causes more and more executions. 

I hope that explains the issue. I'd be more than happy to provide any debug information you might need

**Comment by kevinrademan** [comment]
The public instance is hosted in a kubernetes cluster that is publically accessible. This instance has a webhook trigger and a redis node option in the workflow (i've attached a screenshot)
![image](https://user-images.githubusercontent.com/6704740/195586179-cd1b62ef-c493-4c44-a61c-2c072c7ea7ec.png)

The other instance is hosted in docker desktops kubernetes locally and uses a Redis Trigger and a http request module.
![image](https://user-images.githubusercontent.com/6704740/195586325-f47f9247-8c62-4208-a622-d417224b64f2.png)

The public one works exactly as expected, 1 webhook results in 1 execution of the redis node. Locally however when that redis message comes in the workflow is executed arbitrary amount of times that seems to increase over time.

**Comment by kevinrademan** [comment]
@Joffcom  Here is some screenshots of the execution. I created this simple flow which allows me publish 1 message to redis
![image](https://user-images.githubusercontent.com/6704740/195592875-d0234442-5a51-4cc3-835e-d187375bb92c.png)

And here is an example of what happens locally when I run that workflow (this was after a restart of the pod, more information around that below)
![image](https://user-images.githubusercontent.com/6704740/195594128-a420b4ef-8d14-4d7a-9506-3a1ef14b880d.png)


Finally, restarting the pod running n8n temporarily "fixes" the issue. However over time it starts to increase the number of executions again. Looking at the code I get the feeling there might be some edge cases resuling in the closeFunction not getting called. That would cause the behavior im seeing.
https://github.com/n8n-io/n8n/blob/91d7e16c81675ec1573196271c79cb67503f2ab5/packages/nodes-base/nodes/Redis/RedisTrigger.node.ts#L127

The other thing I've noticed is that locally when I pull the logs I repetely see the  "Loading configuration overwrites from" logs being written. 
![image](https://user-images.githubusercontent.com/6704740/195594027-c3c5132c-412d-441e-b151-6e588ef53cef.png)

Perhaps some kind of soft restart isnt cleaning up the redis client correctly? ITs worth noting that the pod itself isn't restarting, so would have to be something inside the docker container.

**Comment by kevinrademan** [comment]
I enabled debug logging, but no useful information there. Noticed that the Loading conversation logs happen when the sub processes start up so its not a restart causing the leak. The only logs that happened was the health check and this one which doesn't seem related. 

```
2022-10-13T12:57:19.443Z | debug    | Wait tracker querying database for waiting executions "{ file: 'WaitTracker.js', function: 'getWaitingExecutions' }
```

**Comment by Joffcom** [comment]
Hey @kevinrademan,

Looking at this I don't believe it is related to the MQTT trigger issue, I take it the public instance is not using a Redis trigger at all and is just posting the data so the local instance can pick it up?

How many n8n instances are running in the local instance? I am wondering if maybe there are multiple workers and they are picking up the same message.

**Comment by kevinrademan** [comment]
Hey @Joffcom 

_Looking at this I don't believe it is related to the MQTT trigger issue_
Yes I agree. Should I create a new issue with the bug report?

_I take it the public instance is not using a Redis trigger at all and is just posting the data so the local instance can pick it up?_
Correct

_How many n8n instances are running in the local instance? I am wondering if maybe there are multiple workers and they are picking up the same message._
Its a single instance. It serves as just a webhook relay, so I dont really need anything more than that.

**Comment by Joffcom** [comment]
Hey @kevinrademan,

Another issue may be needed or possibly a support post on our community forum as it will get more eyes on it. If it turns out to be a bug and not an odd configuration issue we can then open a dev ticket to fix it.

It may also be worth not using the helm option locally and just using a simple docker setup to see if that changes anything.

**Comment by rvalle** [comment]
@kevinrademan  what % of free space do you have on the database volume?

Could this trigger this kind of behavior? i.e. is there any database queuing in place or similar?

My instability in n8n was growing, not only MQTT, but also some Grafana alerts (webhook) started acting up, with missing text in notifications, etc.

I noticed I had only 15-20% free space in the DB volume, and I remembered at least of a couple DB engines that refuse to process transactions after a certain low mark...

I am still following if this may or may not be the cause, will keep you posted.

**Comment by kevinrademan** [comment]
Hi Guys, sorry for the radio silence, I was on holiday for a week. 

@Joffcom I'll try using docker desktop directly and see how it goes. If I have further issues  I'll switch over to the community.

@rvalle In my case I have plenty of diskspace available (~80%). I also have a few local databases running from this instance and its only the n8n stuff acting up.

**Comment by rvalle** [comment]
@kevinrademan yes... I am still getting the double messages. So, none of the following worked:

- QOS=2 
- Non constrained space on DB volume

@Joffcom I can confirm that non of the previous resolves the issue.

I think the logical step is to start the container with the possibility to connect in about a week for a debug session, and then step into the container when the issue is taking place. I will try to find the deeper point in the stack in which the double message happens.

**Comment by Joffcom** [comment]
Hey @rvalle,

That is interesting we were expecting a QOS of 2 to work 🤔

@csuermann do you have any other thoughts on this one? It seems to be tricky to reproduce with the volume.

**Comment by kevinrademan** [comment]
@Joffcom Just an FYI, today I downloaded the latest n8n desktop app. It does exactly the same thing as my helm chart installation. I think this at the very least rules out configuration being the issue (unless the desktop app has the same configuration issue :p)

**Comment by ulich** [comment]
I am experiencing the same. Interestingly, if i disable the workflow, I still see new executions and then there it's only one execution, no duplicates. Once I enable the workflow again, i receive 2 again. Sometimes disabling and enabling the workflow helps, sometimes it doesn't.

**Comment by macfr76** [comment]
Also have had the MQTT issue for last 2 years, and clearly it is a limitation for having multiple workflow with MQTT Trigger.

Nevertheless I made my flows working using Queue execution mode, that avoids crash of my n8n instance. But it's not very confortable.

Would be great to have a fix.

I had customized the client name in MQTT credentials. I did put it to blak, but it does not seems to change anything. I use Mosquitto as MQTT Server.

**Comment by macfr76** [comment]
Finally not sure, making client name to blank seems to make things better. Will wait till tomorrow and look at logs again. If so, a disclaimer should be inserted in the credentials config UI as suggested by @rvalle last year.

**Comment by rayanpiro** [comment]
> I am experiencing the same. Interestingly, if i disable the workflow, I still see new executions and then there it's only one execution, no duplicates. Once I enable the workflow again, i receive 2 again. Sometimes disabling and enabling the workflow helps, sometimes it doesn't.

Exactly the same for me. Just one workflow taking data and storing in a InfluxDB instance. Sometimes two instances are working at the same time (ms of difference in InfluxDB). If I disable the workflow sometimes both stops and sometimes doesn't. Restarting the container just makes it working again but after some time (hours) it starts to duplicate data again.

No client_id setted in credentials. One dockerized instance with standard config.

**Comment by Joffcom** [comment]
@rayanpiro I really want to get to the bottom of this one, In your set up how many messages are being sent to your MQTT queue / topic and how long does it take for you to start to see the double runs?

Are you using the parallel processing option on the trigger as well?

**Comment by tylerodonnell** [comment]
We are encountering a similar problem with the Redis "new Redis event" trigger, where a single publish event triggers multiple duplicate workflow runs. For instance, there have been instances where one event led to three separate workflow runs. Temporarily disabling and then reenabling the workflow with the 'new Redis event' trigger resolves the issue of duplicate runs. However, the problem resurfaces whenever we modify that workflow.

Our `EXECUTIONS_MODE` is configured as a `queue` and we're running version n8n `v1.21.1`.

**Comment by Joffcom** [comment]
Hey @tylerodonnell,

This is for MQTT not redis, could you create a post on our support forum for the redis issue and we can take a look.

**Comment by tylerodonnell** [comment]
@joffcom can do. I just wanted to add an extra datapoint for consideration, in case it's related to the same underlying issue.

**Comment by Joffcom** [comment]
Hey @tylerodonnell,

It could be but as they are different nodes that use different  node packages it is best to treat them as different issues.

**Comment by nicodemusjls** [comment]
Updated server to verify issue still exists. For me, I see the issue is created only when I use the Execute Flow for testing purposes. I can create a flow from scratch and all works perfectly with one execution per mqtt message. However, if I press the Execute Flow for testing the issue arises. In fact, the longer I leave it in Listening state, the more duplications I'll see. For example, if I click Execute Flow and then immediately turn it back off, then send mqtt message, it will be processed 4-5 times. If I leave it in Listening Mode for 20 seconds before turning it off, I'll end up with some 40 duplicate executes. The duplicate executions continue even after server reboot. Based on the behavior, I suspect that during the Listening Mode, it is establishing subscriptions to the server over and over again. Therefore, the mqtt server is sending the messages those 40 times because it is thinking it has many seperate n8n subscriptions to fulfill. The only way to fix the issue is to clear the mqtt server of all it's subscriptions and then in n8n save the workflow so it will establish a new, single subscription with the mqtt server.

**Comment by Joffcom** [comment]
Hey @nicodemusjls that is perfect, So it is the manual run causing the issue I can work with that.

**Comment by rthomas67** [comment]
I have observed similar behavior to that described by @nicodemusjls on workflows using both the MQTT trigger, **and** RabbitMQ trigger nodes.   In the workflow editor, clicking **Test workflow** or **Test step** (within a trigger node UI panel) creates an additional listener process that does not get stopped or cleaned up when the Stop button on the workflow is clicked, even though the workflow editor makes it **appear** as if the workflow-test and listener have been stopped.  This also seems to happen if the browser is closed or navigated away from the editor while still waiting for a test-trigger event.

*Also, BTW, the "Stop listening" button, within many trigger node types, is unresponsive.  The only way to end those "editor-test-listeners," without using the "Back to canvas" link, seems to be **getting** the trigger input (e.g. actually consuming a matching MQTT message).*

# Theory
**If** this issue actually applies more generally to all n8n trigger types, it would probably be useful to identify when/where/in-what-context a "workflow-editor-test-listener" connection might be getting orphaned without being closed properly (e.g. browser closed w/o stopping a trigger test, stop clicked with no trigger event received, editor navigation abandons a page context with an active, pending test listener, etc.).  If this turns out to be the case, n8n should probably protect itself by setting a short (few minutes, or configurable within limits) timeout on all editor-test-listeners, after which they will be automatically stopped / disconnected / closed without regard to whether the expected test trigger input was received.

# Steps to reproduce:
* Start a standalone n8n instance
  * v1.29.1-community, running as a docker container, with postgres DB (also a docker container). 
* Start a local test instance of RabbitMQ (explanation below ※ re: why MQTT may not be as helpful to monitor or test for orphaned listeners.)
* Create a workflow that uses a RabbitMQ trigger node.
* Leave the RabbitMQ exchange/queue empty (i.e. no messages for the trigger to consume)
* Multiple times (3+), click "Test workflow" followed by Stop button.
* Open RabbitMQ management console and view the Connections tab to see multiple live listener connections from the n8n host, corresponding to **no** active workflows or active workflow-tests in n8n.
  * ※ Mosquitto broker does not appear provide a convenient way to check/count existing connections (best substitute seems to be counting netstat ESTABLISHED connections on port 1883).  RabbitMQ's management UI might make this easier.

# Workaround
So far I have only imagined one mitigation strategy for this, which is to run completely independent instances of n8n for editing and active workflows.

# Test resources
Here's a minimal RabbitMQ docker-compose.yml in case it helps set up a test environment:
* Management UI: http://localhost:15672 - Login: guest/guest - Switch to Connections tab.
* Queue endpoint for trigger: localhost:5672
```
services:
  rabbitmq:
    image: rabbitmq:3.13.0-management
    ports:
      - '5672:5672'
      - '15672:15672'
```

**Comment by craSH** [comment]
Hi - I think I'm having this same issue pretty reliably with the MQTT trigger, set to QoS level 2 for all sides (publisher and n8n trigger subscriber), but I haven't noticed it as aggressively because I built in message de-duplication to my workflow (because my data source is also likely to sent duplicate messages at the app layer, regardless of MQTT QoS aspects). When subscribed to the same topic using a standalone MQTT client, QoS 2, I observe no duplicate messages, but n8n's trigger does indeed fire twice for every message sent to the topic.

If I can provide any diagnostic details to help here, please let me know. Happy to work through it on the n8n discord as well.

**Comment by Joffcom** [comment]
This should now be fixed in 1.48.0 if anyone wants to upgrade and test it would be handy to have some confirmation before we close this.

**Comment by rthomas67** [comment]
I see that this is already closed, but it's the sort of thing that has to be watched for a while to confirm whether it is fixed.  I had a n8n workflow set up to send an MQTT backup request message on a weekly timer, and then respond via MQTT trigger to the resulting message.  BTW, it's really handy that both of those workflow paths can be in the same workflow.  Before upgrading to 1.50.1 (which I assume includes this 1.48.0 update), the trigger from the MQTT backup message was running the workflow around 40 times.   So, not 100% sure yet if the n8n MQTT trigger node is fixed, or if restarting n8n for the upgrade also started over with a single trigger response, but so far so good (only 1 workflow execution from the trigger message for the timer that ran).  Still watching, and will add another comment here if the multi-execution issue returns.

**Comment by craSH** [comment]
I've also not had multiple executions happen for a while now, and have been generally quick to update to the current latest tag.
